### PR TITLE
Optimization of ScaleTransform for SimpleKernel

### PR DESCRIPTION
**Pull request by devmotion**
Fixes https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/96 by adapting the old code to the new interface. I'm not sure, maybe there's a better way to achieve this optimization?

### CODE REVIEW

File: src/kernels/transformedkernel.jl [hunk 1]
```diff
@@ -10,6 +10,19 @@ struct TransformedKernel{Tk<:Kernel,Tr<:Transform} <: Kernel
...
+end
+
+function _scale(t::ScaleTransform, metric::Euclidean, x, y)
+    return first(t.s) * evaluate(metric, x, y)
+end
+function _scale(t::ScaleTransform, metric::Union{SqEuclidean,DotProduct}, x, y)
+    return first(t.s)^2 * evaluate(metric, x, y)
+end
+function _scale(t::ScaleTransform, metric, x, y)
+    evaluate(metric, apply(t, x), apply(t, y))
```
Comment (theogf) [inline src/kernels/transformedkernel.jl:24]: apply is not valid anymore. It should be `t(x)`, `t(y)`.

Comment (devmotion) [reply src/kernels/transformedkernel.jl:24]: Ah true, thanks for spotting it!

**Comment by willtebbutt** [comment]
Thanks for taking a look at this. Before merging, could someone please explain to me why these methods are more efficient than what would otherwise happen? I'm struggling to figure it out.

**Comment by theogf** [comment]
Well if you would multiply `x` by `s` before evaluating the metric you would need to multiply every element of `x` by `s` first. Whereas if you do it afterwards you just do a scalar-scalar multiplication

**Comment by willtebbutt** [comment]
Ah, okay. @devmotion could you add a comment to this effect?

**Comment by devmotion** [comment]
I added a comment and constrained the optimization to `AbstractVector{<:Real}` to be a bit more conservative and avoid surprises for users.

File: src/kernels/transformedkernel.jl [hunk 2]
```diff
@@ -10,7 +10,14 @@ struct TransformedKernel{Tk<:Kernel,Tr<:Transform} <: Kernel
...
 
 (k::TransformedKernel)(x, y) = k.kernel(k.transform(x), k.transform(y))
-function (k::TransformedKernel{<:SimpleKernel,<:ScaleTransform})(x, y)
+
+# Optimizations for scale transforms of simple kernels to save allocations:
+# Instead of a multiplying every element of the inputs before evaluating the metric,
+# we perform a scalar multiplcation of the distance of the original inputs, if possible.
+function (k::TransformedKernel{<:SimpleKernel,<:ScaleTransform})(
+    x::AbstractVector{<:Real},
+    y::AbstractVector{<:Real}
```
Comment (willtebbutt) [inline src/kernels/transformedkernel.jl:19]:
```suggestion
    y::AbstractVector{<:Real},
```

**Pull request merged**